Use requests

bubanoid opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/openprocurement/openprocurement.client.python/41)
<!-- Reviewable:end -->

selurvedu committed:
Support both unicode and bytes in `verify_file`

selurvedu committed:
Add `create_thin_document` for VDR documents

annawzz committed:
Add uploading and patching functions

leits committed:
Merge pull request #40 from kosaniak/master

Add uploading and patching functions

bubanoid committed:
Correct comment for verify_file decorator

requests package as well as restkit package fail to extract correct
file name from the file object derived with open() function. FileIO
object has to be used instead.

bubanoid committed:
Add explicit file closing

Explicit 'file.close()' call was added to verify_file decorator
despite that file is closing at decorator exit. It is done in
accordance with python's zen: "Explicit is better than implicit".

bubanoid committed:
Remove unnecessary kwargs

bubanoid committed:
Make initial request through session

Initial request to get cookies for an instance of 'APIBaseClient`
class was made through 'Session' object.

bubanoid committed:
Unify all string literals

Replace all double quote string defining symbols with single
quote symbols.

bubanoid committed:
Add json parameter to the request function

bubanoid committed:
Introduce new exceptions

Several new exceptions related to http requests were introduced.
These exceptions are raised inside the 'request' function in
accordance with the common rule: "throw early catch late".

bubanoid committed:
Implement Document Service usage

leits commented on openprocurement_client/api_base_client.py:
```diff
@@ -0,0 +1,176 @@
+            return munchify(loads(response_item.text))
+        raise InvalidResponse(response_item)
+
+    def _add_document_in_api(self, url, payload, headers={}):
```
Яка сфера використання цієї функції і чому б там не використати `_create_resource_item`?

leits commented on openprocurement_client/exceptions.py:
```diff
@@ -1,10 +1,122 @@
+        self.status_code \
+            = self.status_code or getattr(response, 'status_code', None)
+        self.response = response
+        Exception.__init__(self)
```
Наскільки пам’ятаю, у класу Exception нічого особливого при ініціалізації не відбувається, тому це можна забрати.
У іншому випадку, тут потрібно було б використовувати super)

leits commented on openprocurement_client/exceptions.py:
```diff
@@ -1,10 +1,122 @@
-class InvalidResponse(Exception):
+"""
+exception classes.
+"""
```
Цей опис зайвий. Назва модуля його самодокоментує)

leits commented on openprocurement_client/document_service_client.py:
```diff
@@ -0,0 +1,83 @@
+    def document_upload(self, url, files=None, headers={}):
+        file_current = files or self.files
+
+        # file_payload = {'file': (file_name, FileIO(file_path, 'rb'))}
```
Чи є необхідність у цих коментований ділянка?

leits commented on openprocurement_client/document_service_client.py:
```diff
@@ -0,0 +1,83 @@
+            auth=self.auth(self.login, self.passwd), params=params
+        )
+
+        self.files = self.files_processing(files)
```
Краще забрати завантаження файлів у момент ініціалізації. Бо виходить доволі неявна конструкція:

```
files = [first_file, second_file]
ds_client = DocumentServiceClient(api_version, resource, files)
ds_client.document_upload()
ds_client.document_upload()
```

Замість того, щоб явно їх завантажити.

leits commented on openprocurement_client/document_service_client.py:
```diff
@@ -0,0 +1,83 @@
+                 params=None):
+
+        host_url_current = host_url or self.host_url
+        super(DocumentServiceClient, self).__init__(
```
При такій ініціалізації з’являються атрибути інстансу для роботи з api.

leits reviewed (commented):
Після виправлень деяких зауважень можна приймати, щоб уже почати тестування ds. 
Але новий клієнт для ds вийшов дуже залежний і пов’язаний із api. 
Бачу потребу перетворення його у повністю незалежну частину для роботи із файлами. Цей незалежний клієнт має інтегруватисяся у клієнт для api як лише один із можливих сервісів зберігання документів.
Якщо я помиляюсь, @ktarasz може мене поправити.

І обов’язково потрібно добавити тести для нового клієнта.

p.s. рев’ю вийшло трохи розгромне, але найважливіша частина - робота із файлами - зроблено добре.

bubanoid commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

_[openprocurement_client/api_base_client.py, line 127 at r8](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KVjaFTZODn5m3m9wWSR-r8-127:-KVyktZslowfovXrt2fl:b-y2yfyk) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/5028f3548efa3ea89d19b996882e24e253a828fc/openprocurement_client/api_base_client.py#L127)):_
<details><summary>

<i>Previously, Leits (Andrew Leitsius) wrote…</i></summary>



> Яка сфера використання цієї функції і чому б там не використати `_create_resource_item`?
> </details>
> Забираю '_add_document_in_api' і використовую '_create_resource_item'.

---

_[openprocurement_client/document_service_client.py, line 31 at r8](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KVjaFTZODn5m3m9wWSS-r8-31:-KVyjV7ZdZ7Qsv74IX8R:btbiam1) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/5028f3548efa3ea89d19b996882e24e253a828fc/openprocurement_client/document_service_client.py#L31)):_
<details><summary>

<i>Previously, Leits (Andrew Leitsius) wrote…</i></summary>



> При такій ініціалізації з’являються атрибути інстансу для роботи з api.
> </details>
> не розумію ((

---

_[openprocurement_client/document_service_client.py, line 36 at r8](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KVjaFTZODn5m3m9wWSS-r8-36:-KVyfA_adx14G7L7Cb3k:b3bo5me) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/5028f3548efa3ea89d19b996882e24e253a828fc/openprocurement_client/document_service_client.py#L36)):_
<details><summary>

<i>Previously, Leits (Andrew Leitsius) wrote…</i></summary>



> Краще забрати завантаження файлів у момент ініціалізації. Бо виходить доволі неявна конструкція:
> 
> ```
> 
> files = [first_file, second_file]
> 
> ds_client = DocumentServiceClient(api_version, resource, files)
> 
> ds_client.document_upload()
> 
> ds_client.document_upload()
> 
> ```
> 
> Замість того, щоб явно їх завантажити.
> </details>
> Згідний. Бачу 2 виходи:
> 1) забрати атрибут 'files' у класу 'DocumentServiceClient' і всюди передавати файли для аплоаду явно
> 2) залишити цей атрибут але поправити метод document_upload у класу DocumentServiceClient таким чином:
> def document_upload(self, url, files, headers={}):
> а на виклику цього метода явно передавати йому files:
>         response = self.ds_client.document_upload(
>             url=url_upload,
>             files=ds_client.files
>             headers=headers
>         )

Як на мене, обидва варіанти мают свої за і проти. Можна піти другим шляхом? В цьому випадку буде мінімальна радакція коду.

---

_[openprocurement_client/document_service_client.py, line 74 at r8](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KVjaFTZODn5m3m9wWSS-r8-74:-KVxwuFbaJopAdaxLLmR:bdvg57z) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/5028f3548efa3ea89d19b996882e24e253a828fc/openprocurement_client/document_service_client.py#L74)):_
<details><summary>

<i>Previously, Leits (Andrew Leitsius) wrote…</i></summary>



> Чи є необхідність у цих коментований ділянка?
> </details>
> Необхідності у цьому коментарі нема. Це пояснення до коду мало на меті дати зрозуміти, яку форму повинна мати змінна file_payload. Чи достатньо буде змінити цей коментар на такий:
> 
> # file_payload variable should be of the following form: file_payload = {'file': (file_name, FileIO(file_path, 'rb'))}
> 
> ?

---

_[openprocurement_client/exceptions.py, line 3 at r8](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KVWUWlFhNPzWQx6agei-r8-3:-KVyhIISxOYJa3zc7S28:bb6un1j) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/5028f3548efa3ea89d19b996882e24e253a828fc/openprocurement_client/exceptions.py#L3)):_
<details><summary>

<i>Previously, Leits (Andrew Leitsius) wrote…</i></summary>



> Цей опис зайвий. Назва модуля його самодокоментує)
> </details>
> Done

---

_[openprocurement_client/exceptions.py, line 16 at r8](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KVWUWlFhNPzWQx6agei-r8-16:-KVyhX5Ics0cSyXGs_2X:b-noroli) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/5028f3548efa3ea89d19b996882e24e253a828fc/openprocurement_client/exceptions.py#L16)):_
<details><summary>

<i>Previously, Leits (Andrew Leitsius) wrote…</i></summary>



> Наскільки пам’ятаю, у класу Exception нічого особливого при ініціалізації не відбувається, тому це можна забрати.
> 
> У іншому випадку, тут потрібно було б використовувати super)
> </details>
> Без проблем, можна і забрати. Я слідував порадам із http://stackoverflow.com/questions/1319615/proper-way-to-declare-custom-exceptions-in-modern-python , де кажуть, що навішати атрибути на свій кастомний ексепшин є нормально і крім того запускають ініт від батьківського класу Exception, але передавати у нього параметр 'message'. Єдине, справді слід використувувати super. Це є адаптація коду з restkit-а. Я пропустив цей косяк.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41)_

<!-- Sent from Reviewable.io -->

bubanoid commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

_[openprocurement_client/exceptions.py, line 16 at r8](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KVWUWlFhNPzWQx6agei-r8-16:-KVzd1mB_N9c2KiPsf-e:bczotws) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/5028f3548efa3ea89d19b996882e24e253a828fc/openprocurement_client/exceptions.py#L16)):_
<details><summary>

<i>Previously, bubanoid (Andriy Andrusyk) wrote…</i></summary>



> Без проблем, можна і забрати. Я слідував порадам із http://stackoverflow.com/questions/1319615/proper-way-to-declare-custom-exceptions-in-modern-python , де кажуть, що навішати атрибути на свій кастомний ексепшин є нормально і крім того запускають ініт від батьківського класу Exception, але передавати у нього параметр 'message'. Єдине, справді слід використувувати super. Це є адаптація коду з restkit-а. Я пропустив цей косяк.
> </details>
> Зрозумів. Подивився у **init** класу Exception і там стоїть pass. Забираю Exception.**init**(self) взагалі.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41)_

<!-- Sent from Reviewable.io -->

bubanoid commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

_[openprocurement_client/exceptions.py, line 16 at r8](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KVWUWlFhNPzWQx6agei-r8-16:-KVze0wsgDJSUq6EsKjB:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/5028f3548efa3ea89d19b996882e24e253a828fc/openprocurement_client/exceptions.py#L16)):_
<details><summary>

<i>Previously, bubanoid (Andriy Andrusyk) wrote…</i></summary>



> Зрозумів. Подивився у **init** класу Exception і там стоїть pass. Забираю Exception.**init**(self) взагалі.
> </details>
> Done.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41)_

<!-- Sent from Reviewable.io -->

ktarasz commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks broke.

---

_[openprocurement_client/api_base_client.py, line 128 at r10](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KW2wW-T1UlEanLxCI3S:-KW2wW-T1UlEanLxCI3T:b-mqu1ns) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/05d06016b8268eba2958625a54d13064f067de02/openprocurement_client/api_base_client.py#L128)):_

> ``` Python
>         raise InvalidResponse(response_item)
> 
>     def _upload_resource_file(self, url, files=None, headers={}):
> ```

ds_client треба передавати при виклику _upload_resource_file

_upload_resource_file(self, url, files=None, headers={}, ds_client=None):
if ds_client :
     - upload to api with document service client
else
     - log deprecation warning
     - upload to api with POST

---

_[openprocurement_client/document_service_client.py, line 36 at r8](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KVjaFTZODn5m3m9wWSS-r8-36:-KW2xrPNmkF7UK5uC-J2:b-7r09z9) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/5028f3548efa3ea89d19b996882e24e253a828fc/openprocurement_client/document_service_client.py#L36)):_

> files
> 1 файл за раз - на рівні REST нема multiple upload

---

_[openprocurement_client/document_service_client.py, line 75 at r10](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KW2yJxzhzBtpR2dEy2f:-KW2yJxzhzBtpR2dEy2g:b-8sed4h) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/05d06016b8268eba2958625a54d13064f067de02/openprocurement_client/document_service_client.py#L75)):_

> коментований
> file_payload = {'file': (file_name, FileIO(file_path, 'rb'))} 

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41)_

<!-- Sent from Reviewable.io -->

bubanoid commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

_[openprocurement_client/api_base_client.py, line 128 at r10](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KW2wW-T1UlEanLxCI3S:-KW7ZrA4Eyp2YDpKH8pE:b7jh0po) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/05d06016b8268eba2958625a54d13064f067de02/openprocurement_client/api_base_client.py#L128)):_
<details><summary>

<i>Previously, ktarasz (Taras Kozlovskyi) wrote…</i></summary>



> ds_client треба передавати при виклику _upload_resource_file
> 
> _upload_resource_file(self, url, files=None, headers={}, ds_client=None):
> if ds_client :
>      - upload to api with document service client
> else
>      - log deprecation warning
>      - upload to api with POST
> </details>
> Як я розумію, ти це кажеш зробити для того, щоб залишалася можливість аплоадити файли НЕ через DS. Зараз код дозволяє аплоадити ТІЛЬКИ через DS. Якщо лише в цьому справа, то я пропоную зробити так:

_upload_resource_file(self, url, files=None, headers={}, use_ds_client=True):
if ds_client :
- upload to api with document service client
  else
- log deprecation warning
- upload to api with POST

В такому випадку не доведеться передавати всюди параметр ds_client у всіх ф-ціях, що викликають _upload_resource_file. У мене ds_client вже є атрибутом самого клієнта. Чи справді нам потрібен функціонал, який передбачає наявність різних ds_client -ів?

---

_[openprocurement_client/document_service_client.py, line 75 at r10](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KW2yJxzhzBtpR2dEy2f:-KW5BDqoMAhHXoWZoxvG:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/05d06016b8268eba2958625a54d13064f067de02/openprocurement_client/document_service_client.py#L75)):_
<details><summary>

<i>Previously, ktarasz (Taras Kozlovskyi) wrote…</i></summary>



> > коментований
> > file_payload = {'file': (file_name, FileIO(file_path, 'rb'))} 
> > </details>
> > Done.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41)_

<!-- Sent from Reviewable.io -->

bubanoid commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

_[openprocurement_client/api_base_client.py, line 128 at r10](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KW2wW-T1UlEanLxCI3S:-KW7bUcDbSSni0jOHL6h:b-q3kxf7) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/05d06016b8268eba2958625a54d13064f067de02/openprocurement_client/api_base_client.py#L128)):_
<details><summary>

<i>Previously, bubanoid (Andriy Andrusyk) wrote…</i></summary>



> Як я розумію, ти це кажеш зробити для того, щоб залишалася можливість аплоадити файли НЕ через DS. Зараз код дозволяє аплоадити ТІЛЬКИ через DS. Якщо лише в цьому справа, то я пропоную зробити так:
> 
> _upload_resource_file(self, url, files=None, headers={}, use_ds_client=True):
> if ds_client :
> - upload to api with document service client
>   else
> - log deprecation warning
> - upload to api with POST
> 
> В такому випадку не доведеться передавати всюди параметр ds_client у всіх ф-ціях, що викликають _upload_resource_file. У мене ds_client вже є атрибутом самого клієнта. Чи справді нам потрібен функціонал, який передбачає наявність різних ds_client -ів?
> </details>
> По дурнуватому відформатувало код. Я ще не вмію нормально тут форматувати. Мало бути так:

``` python
_upload_resource_file(self, url, files=None, headers={}, use_ds_client=True):
    if ds_client:
        - upload to api with document service client 
    else
        - log deprecation warning
        - upload to api with POST
```

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41)_

<!-- Sent from Reviewable.io -->

bubanoid commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

_[openprocurement_client/api_base_client.py, line 128 at r10](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KW2wW-T1UlEanLxCI3S:-KW7c5GGCaLstfzSoHIu:buzyzrk) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/05d06016b8268eba2958625a54d13064f067de02/openprocurement_client/api_base_client.py#L128)):_
<details><summary>

<i>Previously, bubanoid (Andriy Andrusyk) wrote…</i></summary>



> По дурнуватому відформатувало код. Я ще не вмію нормально тут форматувати. Мало бути так:
> 
> _upload_resource_file(self, url, files=None, headers={}, use_ds_client=True):
> if ds_client :
> - upload to api with document service client 
>   else
> - log deprecation warning
> - upload to api with POST
> 
> </details>
> Знову зле. Ти код і так зрозумієш.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41)_

<!-- Sent from Reviewable.io -->

bubanoid commented:
Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

_[openprocurement_client/api_base_client.py, line 128 at r10](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KW2wW-T1UlEanLxCI3S:-KWjflrr7vT1VbPC0_6_:b-zg1gm7) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/05d06016b8268eba2958625a54d13064f067de02/openprocurement_client/api_base_client.py#L128)):_
<details><summary>

<i>Previously, bubanoid (Andriy Andrusyk) wrote…</i></summary>



> Знову зле. Ти код і так зрозумієш.
> </details>
> Зробив так як треба, тобто ds_client передається при виклику _upload_resource_file. Отже, done.

---

_[openprocurement_client/document_service_client.py, line 31 at r8](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KVjaFTZODn5m3m9wWSS-r8-31:-KWjfDMMu7D3rPNaJMvu:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/5028f3548efa3ea89d19b996882e24e253a828fc/openprocurement_client/document_service_client.py#L31)):_
<details><summary>

<i>Previously, bubanoid (Andriy Andrusyk) wrote…</i></summary>



> не розумію ((
> </details>
> Done.

---

_[openprocurement_client/document_service_client.py, line 36 at r8](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KVjaFTZODn5m3m9wWSS-r8-36:-KWjfGWeVk1taHM76DfQ:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/5028f3548efa3ea89d19b996882e24e253a828fc/openprocurement_client/document_service_client.py#L36)):_
<details><summary>

<i>Previously, ktarasz (Taras Kozlovskyi) wrote…</i></summary>



> > files
> > 1 файл за раз - на рівні REST нема multiple upload
> > </details>
> > Done.

---

_[openprocurement_client/document_service_client.py, line 74 at r8](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KVjaFTZODn5m3m9wWSS-r8-74:-KWjfLEhL-8jqWS78zZS:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/5028f3548efa3ea89d19b996882e24e253a828fc/openprocurement_client/document_service_client.py#L74)):_
<details><summary>

<i>Previously, bubanoid (Andriy Andrusyk) wrote…</i></summary>



> Необхідності у цьому коментарі нема. Це пояснення до коду мало на меті дати зрозуміти, яку форму повинна мати змінна file_payload. Чи достатньо буде змінити цей коментар на такий:
> 
> # file_payload variable should be of the following form: file_payload = {'file': (file_name, FileIO(file_path, 'rb'))}
> 
> ?
> </details>
> Done.

---

_[openprocurement_client/document_service_client.py, line 75 at r10](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KW2yJxzhzBtpR2dEy2f:-KWjgWYuKxtCaRnj6J5p:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/05d06016b8268eba2958625a54d13064f067de02/openprocurement_client/document_service_client.py#L75)):_
<details><summary>

<i>Previously, bubanoid (Andriy Andrusyk) wrote…</i></summary>



> Done.
> </details>
> Done.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41)_

<!-- Sent from Reviewable.io -->

bubanoid committed:
Add ds_client argument

ds_client argument was added to all methods connected to file
operations. The default value of ds_client is None that allows
file operations (download/upload/delete) performend with DS or
without it.

bubanoid committed:
Fix bug with different form of response header

Amazon DS returns http response with two different formats of
Content-Disposition header. Both formats must be supported.

bubanoid committed:
Set correct status codes of http response

Expected status codes for successful 'put', 'delete' methods of http
requests are 200 or 204, not 201.

bubanoid committed:
Fix tests to fit changes in Client

bubanoid committed:
Fix upload_document function in ContractingClient

bubanoid committed:
Add separate PORT constant to test.py module

bubanoid committed:
Improve imports in tests.py

bubanoid committed:
Introduce setting_up function

VDigitall committed:
Fix sync.py

bubanoid committed:
Handle start Bottle server exceptions

bubanoid committed:
Introduce BaseTestClass class

bubanoid committed:
Fix unittests to test file operations with DS

bubanoid committed:
Introduce _get_access_token function

bubanoid committed:
Improve client work with DS

DS-client instance is constructed outside of the client constructor
and is passed as an argument to the client constructor. Then it is
saved as one of the client attributes.

bubanoid committed:
Merge pull request #1 from VDigitall/use_requests_fix

Fix sync.py and api_base_client.py

selurvedu committed:
Merge pull request #42 from bubanoid/filename_token_format

Fix bug with different form of response header

bubanoid committed:
Merge branch 'master' into use_requests

bubanoid committed:
Fix conflicts

Merge remote-tracking branch 'openprocurement/use_requests' into
use_requests. User-Agent header added.

ktarasz commented:
This branch has conflicts that must be resolved

ktarasz commented:
This branch has conflicts that must be resolved

bubanoid committed:
Refactoring

Function '_patch_obj_resource_item' was introduced and placed into the
'APIBaseClient' class. Method 'upload_document' was moved from
'TendersClient' and 'ContractingClient' classes to parent class.

bubanoid committed:
Fix conflicts

Merge remote-tracking branch 'openprocurement/use_requests' into
use_requests.

bubanoid commented:
Conflicts are fixed

bubanoid committed:
Fix conflicts

Merge remote-tracking branch 'openprocurement/use_requests' into
use_requests

bubanoid committed:
Rename variable 'files' to 'file_'

bubanoid committed:
Remove default values for client in utils.py

bubanoid committed:
Refactor work with Document Service

bubanoid committed:
Extend DS-client functionality

The possibility to upload file through DS without preregistration was
added.

bubanoid committed:
Improve handling DS register url and upload url

selurvedu commented:
Переглянув частину файлів, це ще не все.

---

Reviewed 1 of 4 files at r10, 5 of 10 files at r11.
Review status: 6 of 11 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[openprocurement_client/sync.py, line 45 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYUWthpK_MsJMj0y8_1:-KYUWthpK_MsJMj0y8_2:b-psrlb4) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/sync.py#L45)):*
(Skip)

> backfard

backward

---

*[openprocurement_client/tests/_server.py, line 24 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYUXLaIjuLNsdqiSKgF:-KYUXLaJWwjxyDuYQa99:b-hsxu94) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/tests/_server.py#L24)):*
(Skip)

> routs

routes

---

*[openprocurement_client/tests/_server.py, line 317 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYUYDdoJjSrFamC0qdn:-KYUYDdpSQB_xmVNh7sN:b66gxev) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/tests/_server.py#L317)):*
> ```Python
> 
> 
> download_url_extension = 'some_key_etc'
> ```

Це константа, і її варто винести на початок файлу і написати в uppercase. Заодно перевір, чи є ще десь таке.

Thx to @ktarasz.

---

*[openprocurement_client/tests/_server.py, line 333 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYUYODnk8IoKHfAKHOC:-KYUYODnk8IoKHfAKHOD:b-l0zrsh) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/tests/_server.py#L333)):*
> ```Python
> 
> 
> routs_dict_ds = {
> ```

routes

---

*[openprocurement_client/tests/tests.py, line 23 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYUZnb4pwj4WZgGFZBD:-KYUZnb5C1queBIr1GXX:b-bpoer4) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/tests/tests.py#L23)):*
> ```Python
> 
> import logging
> logging.basicConfig()
> ```

Я не бачу використання logging у цьому файлі.

---

*[openprocurement_client/tests/tests.py, line 73 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYU_MqvFPQSbgVOh44x:-KYU_Mqw3IjwKxqFM5bG:b-tdo2lw) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/tests/tests.py#L73)):*
> ```Python
>             self.server.start()
>         except Exception as error:
>             print(sys.exc_info()[0], sys.exc_info()[1], sys.exc_info()[2])
> ```

Додай або print в stderr, або в логер, налаштований на stderr, наприклад:
```python
from __future__ import print_function

print(sys.exc_info()[0], sys.exc_info()[1], sys.exc_info()[2], file=sys.stderr)
```

---

*[openprocurement_client/tests/tests.py, line 87 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYUatKTd5OYcFIGh_2N:-KYUatKUik3AbWjt-PoW:bxj5hxd) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/tests/tests.py#L87)):*
> ```Python
>             cls.server_ds.start()
>         except Exception as error:
>             print(sys.exc_info()[0], sys.exc_info()[1], sys.exc_info()[2])
> ```

Тут так само з print.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-:-KYUdlpXws5Osl8gZ1pZ:bivsr08)*
<!-- Sent from Reviewable.io -->

leits commented:
Travis CI не пройшов

bubanoid committed:
Improve unittests

bubanoid commented:
Review status: 4 of 11 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[openprocurement_client/sync.py, line 45 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYUWthpK_MsJMj0y8_1:-KYnsmTY9yP_hZis9cyl:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/sync.py#L45)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

(Skip)

> backfard

backward
</blockquote></details>

Done.

---

*[openprocurement_client/tests/_server.py, line 24 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYUXLaIjuLNsdqiSKgF:-KYnsT6DB5guYV0R8c28:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/tests/_server.py#L24)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

(Skip)

> routs

routes
</blockquote></details>

Done.

---

*[openprocurement_client/tests/_server.py, line 317 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYUYDdoJjSrFamC0qdn:-KYntG3702QWCbvolsq2:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/tests/_server.py#L317)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

Це константа, і її варто винести на початок файлу і написати в uppercase. Заодно перевір, чи є ще десь таке.

Thx to @ktarasz.
</blockquote></details>

Done.

---

*[openprocurement_client/tests/_server.py, line 333 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYUYODnk8IoKHfAKHOC:-KYnsLW4-sjdV_qdfumC:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/tests/_server.py#L333)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

routes
</blockquote></details>

Done.

---

*[openprocurement_client/tests/tests.py, line 23 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYUZnb4pwj4WZgGFZBD:-KYnhXQsaoq4fbmj63GN:b-vahfs2) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/tests/tests.py#L23)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

Я не бачу використання logging у цьому файлі.
</blockquote></details>

logging потрібен на випадок, якщо треба запустити unittest в якому файлові операції виконуться без DS. В цьому випадку клас **APIBaseClient** генерує warning в якому повідомляється, що "**File upload/download/delete outside of the Document Service is deprecated**". Я помістив цей шматок коду у коментар:
```python
cls.ds_client = DocumentServiceClient(host_url=DS_HOST_URL, auth_ds=AUTH_DS_FAKE)
# to test units performing file operations outside the DS uncomment
# following lines:
# import logging
# logging.basicConfig()
# cls.ds_client = None
```
Слід було би зробити так, щоб код проганявся автоматично в двох варіантах: із використанням DS і без нього. Потім, коли робота із файлами поза DS буде закрита, то варіант без DS був би відрізаний. Я знаю як це гарно зробити у **pytest** (використовуючи парметризовані фікстури), але, на жаль, не знаю, як це гарно зробити у **unittest**. Тому пропоную на разі залишити цей кривенький варіант, коли при неообхідності розкоментується стрічка `cls.ds_client = None` і тоді тест пробіжить без використання DS.

---

*[openprocurement_client/tests/tests.py, line 73 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYU_MqvFPQSbgVOh44x:-KYnqdayGgRBQlzp1ZyB:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/tests/tests.py#L73)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

Додай або print в stderr, або в логер, налаштований на stderr, наприклад:
```python
from __future__ import print_function

print(sys.exc_info()[0], sys.exc_info()[1], sys.exc_info()[2], file=sys.stderr)
```
</blockquote></details>

Done.

---

*[openprocurement_client/tests/tests.py, line 87 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYUatKTd5OYcFIGh_2N:-KYnqc1eFf1DMGoS2NAc:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/tests/tests.py#L87)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

Тут так само з print.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41)*
<!-- Sent from Reviewable.io -->

bubanoid committed:
Minor

selurvedu commented:
Almost done! :tada:

---

Reviewed 5 of 10 files at r11, 4 of 4 files at r12, 3 of 3 files at r13.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[openprocurement_client/api_base_client.py, line 61 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoNRltuuZZL31XB4wt:-KYoNRluBvc8F0crPdiG:bvy22bv) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/api_base_client.py#L61)):*
> ```Python
> 
>     @staticmethod
>     def auth(login, passwd): return HTTPBasicAuth(login, passwd)
> ```

Isn't this useless?

`from requests.auth import HTTPBasicAuth as BasicAuth` чи `Auth`.

---

*[openprocurement_client/api_base_client.py, line 89 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoPICcVKt5y2aehG2p:-KYoPICdy7eOWDWEebLu:b-493thj) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/api_base_client.py#L89)):*
> ```Python
> 
>         if response.status_code >= 400:
>             if response.status_code == 404:
> ```

BTW
Давай ти в окремому PR зробиш dict, де ключем буде status code, а значенням – відповідний exception (клас), і це буде в exceptions.py.

---

*[openprocurement_client/api_base_client.py, line 186 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoU0j2GJ1w4F3y2O29:-KYoU0j3VeDXHqn7bB_w:b-h37vs8) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/api_base_client.py#L186)):*
> ```Python
>         _headers.update(headers or {})
>         response_item = self.request(
>             'PATCH', url, headers=_headers, payload=dumps(payload)
> ```

Можна замінити `payload=dumps(payload)` на `json=payload` і забрати dumps з імпортів (loads залишиться).

---

*[openprocurement_client/api_base_client.py, line 239 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoS32-9-4CnSjbSvdS:-KYoS32-9-4CnSjbSvdT:bjv2xfi) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/api_base_client.py#L239)):*
> item_obj, items_name

Хотілося би щоб спочатку йшло ім'я об'єкта, а потім його значення. Перевір, чи можна так.
І  в назві одного`item`, а в іншого – `items`.

---

*[openprocurement_client/api_base_client.py, line 266 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoSqARXhPs_W4rqpIw:-KYoSqATD1eZY0Ya9buu:b-8yvc5x) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/api_base_client.py#L266)):*
> ```Python
>         )
> 
>     def get_resource_item(self, id):
> ```

Будь ласка, додай `headers=None`.

---

*[openprocurement_client/contract.py, line 1 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoYiZmCsxtgVH4xvDo:-KYoYiZnaQEftGyXzfxr:b-879xxq) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/contract.py#L1)):*
> ```Python
> from json import loads
> ```

Заміни json на simplejson, будь ласка.

---

*[openprocurement_client/document_service_client.py, line 15 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoUVuKpOe-QJEgoAe9:-KYoUVuKpOe-QJEgoAeA:b9g9iw5) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/document_service_client.py#L15)):*
> ```Python
> 
> class DocumentServiceClient(APITemplateClient):
>     """base class for API"""
> ```

Некоректний docstring.

---

*[openprocurement_client/document_service_client.py, line 79 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoXH3BXVSWd6FXLNpO:-KYoXH3CdFS8TmvS03C8:b-8rb4zr) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/document_service_client.py#L79)):*
> ```Python
>         file_info_dict = self.file_info(file_)
> 
>         response_reg = self.register_document_upload(
> ```

s/response_reg/response/g

---

*[openprocurement_client/exceptions.py, line 107 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoKBCcX9lylADuxaVi:-KYoKBCdCaSg6NgGPw3M:bmzzvc7) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/exceptions.py#L107)):*
> ```Python
> 
> 
> class InvalidResponse(ResourceError):
> ```

Перед цим класом варто би вліпити якийсь коментар про те, що цей exception і наступні є специфічними для `openprocurement.client`, і такий самий коментар для тих, що вище, про те, що це ексепшни для HTTP status codes.

---

*[openprocurement_client/tests/tests.py, line 23 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYUZnb4pwj4WZgGFZBD:-KYoeQcQlieazgWYK3qk:b-jbsf8z) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/tests/tests.py#L23)):*
<details><summary><i>Previously, bubanoid (Andriy Andrusyk) wrote…</i></summary><blockquote>

logging потрібен на випадок, якщо треба запустити unittest в якому файлові операції виконуться без DS. В цьому випадку клас **APIBaseClient** генерує warning в якому повідомляється, що "**File upload/download/delete outside of the Document Service is deprecated**". Я помістив цей шматок коду у коментар:
```python
cls.ds_client = DocumentServiceClient(host_url=DS_HOST_URL, auth_ds=AUTH_DS_FAKE)
# to test units performing file operations outside the DS uncomment
# following lines:
# import logging
# logging.basicConfig()
# cls.ds_client = None
```
Слід було би зробити так, щоб код проганявся автоматично в двох варіантах: із використанням DS і без нього. Потім, коли робота із файлами поза DS буде закрита, то варіант без DS був би відрізаний. Я знаю як це гарно зробити у **pytest** (використовуючи парметризовані фікстури), але, на жаль, не знаю, як це гарно зробити у **unittest**. Тому пропоную на разі залишити цей кривенький варіант, коли при неообхідності розкоментується стрічка `cls.ds_client = None` і тоді тест пробіжить без використання DS.
</blockquote></details>

<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

*[openprocurement_client/tests/tests.py, line 69 at r12](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYobJNwZZhxiogiv1j8:-KYobJNxqr_SQZRh3K9P:b-34syhc) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/8a264ad07aad6208c93f5791aaf23a4881b486e7/openprocurement_client/tests/tests.py#L69)):*
> ```Python
>     file_.write(content)
>     file_.seek(0)
>     return munchify(file_)
> ```

Тут munchify зайвий.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-:-KYofDin65Qw9yzjNgbw:b4pzx5w)*
<!-- Sent from Reviewable.io -->

bubanoid committed:
Delete unnecessary method 'auth'

bubanoid committed:
Introduce dict with http exceptions

bubanoid committed:
Minor improvements

bubanoid commented:
Review status: 5 of 11 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[openprocurement_client/api_base_client.py, line 61 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoNRltuuZZL31XB4wt:-KYrn6P39RSbpm82DIkF:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/api_base_client.py#L61)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

Isn't this useless?

`from requests.auth import HTTPBasicAuth as BasicAuth` чи `Auth`.
</blockquote></details>

Done.

---

*[openprocurement_client/api_base_client.py, line 89 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoPICcVKt5y2aehG2p:-KYrvNXX1Ej63xEDFMhD:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/api_base_client.py#L89)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

BTW
Давай ти в окремому PR зробиш dict, де ключем буде status code, а значенням – відповідний exception (клас), і це буде в exceptions.py.
</blockquote></details>

Done.

---

*[openprocurement_client/api_base_client.py, line 186 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoU0j2GJ1w4F3y2O29:-KYrwU5rX-um8aKaBzAU:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/api_base_client.py#L186)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

Можна замінити `payload=dumps(payload)` на `json=payload` і забрати dumps з імпортів (loads залишиться).
</blockquote></details>

Done.

---

*[openprocurement_client/api_base_client.py, line 239 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoS32-9-4CnSjbSvdS:-KYrzptRKs0A_ZHLc1yu:budorf6) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/api_base_client.py#L239)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

> item_obj, items_name

Хотілося би щоб спочатку йшло ім'я об'єкта, а потім його значення. Перевір, чи можна так.
І  в назві одного`item`, а в іншого – `items`.
</blockquote></details>

Власне, `items_name` не є назвою об'єкта. Це лише підстрічка в url до api, на який буде відправлятися запит. Тому мінти порядок змінних не бачу сенсу. Крім того, у нас `items_name` приймає значення ``bids`, `awards`, `documents`, `lots`, `contracts`. Тому і `items_name`, а не `item_name`. У цей пул-реквест я міняти не буду, і Тарас чи Андрій зараз пул-реквест приймуть. Але якщо ти вважаєш, що таки треба зробити, як ти кажеш, то роби задачу в асані, я поправлю.

---

*[openprocurement_client/api_base_client.py, line 266 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoSqARXhPs_W4rqpIw:-KYrxKw5hl9XpmnkxRAH:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/api_base_client.py#L266)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

Будь ласка, додай `headers=None`.
</blockquote></details>

Done.

---

*[openprocurement_client/contract.py, line 1 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoYiZmCsxtgVH4xvDo:-KYrxyGNUf4pz2BFpJGY:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/contract.py#L1)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

Заміни json на simplejson, будь ласка.
</blockquote></details>

Done.

---

*[openprocurement_client/document_service_client.py, line 15 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoUVuKpOe-QJEgoAe9:-KYs1BCeNVT37v_nnrKP:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/document_service_client.py#L15)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

Некоректний docstring.
</blockquote></details>

Done.

---

*[openprocurement_client/document_service_client.py, line 79 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoXH3BXVSWd6FXLNpO:-KYs1RrmQ6-THx3dsWQV:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/document_service_client.py#L79)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

s/response_reg/response/g
</blockquote></details>

Done.

---

*[openprocurement_client/exceptions.py, line 107 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoKBCcX9lylADuxaVi:-KYs1WKHdmWZCsQ6a42r:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/exceptions.py#L107)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

Перед цим класом варто би вліпити якийсь коментар про те, що цей exception і наступні є специфічними для `openprocurement.client`, і такий самий коментар для тих, що вище, про те, що це ексепшни для HTTP status codes.
</blockquote></details>

Done.

---

*[openprocurement_client/tests/tests.py, line 69 at r12](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYobJNwZZhxiogiv1j8:-KYs1i185q0k3OAQq8ra:b-896fix) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/8a264ad07aad6208c93f5791aaf23a4881b486e7/openprocurement_client/tests/tests.py#L69)):*
<details><summary><i>Previously, selurvedu wrote…</i></summary><blockquote>

Тут munchify зайвий.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41)*
<!-- Sent from Reviewable.io -->

selurvedu commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r14, 2 of 2 files at r15, 5 of 5 files at r16.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[openprocurement_client/api_base_client.py, line 239 at r11](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-KYoS32-9-4CnSjbSvdS:-KZWV6YPM6bZCChWGecZ:b728gy1) ([raw file](https://github.com/openprocurement/openprocurement.client.python/blob/c917d8530b1e2865db8cca61bcb97cb58721356b/openprocurement_client/api_base_client.py#L239)):*
<details><summary><i>Previously, bubanoid (Andriy Andrusyk) wrote…</i></summary><blockquote>

Власне, `items_name` не є назвою об'єкта. Це лише підстрічка в url до api, на який буде відправлятися запит. Тому мінти порядок змінних не бачу сенсу. Крім того, у нас `items_name` приймає значення ``bids`, `awards`, `documents`, `lots`, `contracts`. Тому і `items_name`, а не `item_name`. У цей пул-реквест я міняти не буду, і Тарас чи Андрій зараз пул-реквест приймуть. Але якщо ти вважаєш, що таки треба зробити, як ти кажеш, то роби задачу в асані, я поправлю.
</blockquote></details>

Я мав на увазі наступне:

Мене здивувало те, що в URL запиту спочатку йде items_name і  потім id, взяте із items_obj, а в аргументах методу спочатку йде items_obj, а потім items_name. Задачу виписав, але вона не є важливою чи високопріоритетною. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openprocurement/openprocurement.client.python/41#-:-KZWV2N17v0HfNM4dUJo:b-jbsf8z)*
<!-- Sent from Reviewable.io -->
